tests/dash/gnmi_utils.py

```diff
@@ -0,0 +1,377 @@
+            pb.vnet_direct.vnet = json_obj["vnet"]
+            pb.vnet_direct.overlay_ip.ipv4 = socket.htonl(int(ipaddress.IPv4Address(json_obj["overlay_ip"])))
+        elif json_obj["action_type"] == "direct":
+            pb.action_type = RoutingType.ROUTING_TYPE_DIRECT
+        else:
+            pytest.fail("Unknown action type %s" % json_obj["action_type"])
+    elif table_name == "ROUTE_RULE":
+        pb = RouteRule()
+        pb.action_type = RoutingType.ROUTING_TYPE_VNET_ENCAP
+        pb.priority = int(json_obj["priority"])
+        pb.pa_validation = json_obj["pa_validation"] == "true"
+        pb.vnet = json_obj["vnet"]
+    elif table_name == "ROUTING_TYPE":
+        pb = RouteType()
+        pbi = RouteTypeItem()
+        pbi.action_name = json_obj["name"]
+        pbi.action_type = ActionType.ACTION_TYPE_MAPROUTING
+        pb.items.append(pbi)
+    else:
+        pytest.fail("Unknown table %s" % table_name)
+    return pb.SerializeToString()
+
+
+def apply_gnmi_file(duthost, ptfhost, dest_path):
+    env = GNMIEnvironment(duthost)
+    logger.info("Applying config files on DUT")
+    dut_command = "cat %s" % dest_path
+    ret = duthost.shell(dut_command)
+    assert ret["rc"] == 0, "Failed to read config file"
+    text = ret["stdout"]
+    res = json.loads(text)
+    delete_list = []
+    update_list = []
+    update_cnt = 0
+    for operation in res:
+        if operation["OP"] == "SET":
+            for k, v in operation.items():
+                if k == "OP":
+                    continue
+                logger.info("Config Json %s" % k)
+                update_cnt += 1
+                filename = "update%u" % update_cnt
+                if ENABLE_PROTO:
+                    message = json_to_proto(k, v)
+                    with open(env.work_dir+filename, "wb") as file:
+                        file.write(message)
+                else:
+                    text = json.dumps(v)
+                    with open(env.work_dir+filename, "w") as file:
+                        file.write(text)
+                ptfhost.copy(src=env.work_dir+filename, dest='/root/')
+                keys = k.split(":", 1)
+                k = keys[0] + "[key=" + keys[1] + "]"
+                if ENABLE_PROTO:
+                    path = "/APPL_DB/%s:$/root/%s" % (k, filename)
+                else:
+                    path = "/APPL_DB/%s:@/root/%s" % (k, filename)
+                update_list.append(path)
+        elif operation["OP"] == "DEL":
+            for k, v in operation.items():
+                if k == "OP":
+                    continue
+                keys = k.split(":", 1)
+                k = keys[0] + "[key=" + keys[1] + "]"
+                path = "/APPL_DB/%s" % (k)
+                delete_list.append(path)
+        else:
+            logger.info("Invalid operation %s" % operation["OP"])
+    ret, msg = gnmi_set(duthost, ptfhost, delete_list, update_list, [])
+    assert ret == 0, msg
+    time.sleep(5)
```

Comment:
Why do we need to sleep here?

Reply:
GNMI will update APPL_DB, and orchagent can't invoke SAI immediately.
We have to wait a few seconds to verify dataplane.

Reply:
In the future, we will have a feedback loop so that we don't need to sleep but need to wait for a dedicated event